xdsl/dialects/hw.py

```diff
@@ -409,12 +426,443 @@ def print_parameters(self, printer: Printer):
             printer.print_string("]")
 
 
+class Direction(Enum):
+    """
+    Represents the direction of a module port.
+    """
+
+    # TODO: support INOUT direction (https://github.com/xdslproject/xdsl/issues/2368)
+    INPUT = (0,)
+    OUTPUT = (1,)
+
+    @staticmethod
+    def parse_optional(parser: BaseParser, short: bool = False) -> "Direction | None":
+        if parser.parse_optional_keyword("input" if not short else "in"):
+            return Direction.INPUT
+        elif parser.parse_optional_keyword("output" if not short else "out"):
+            return Direction.OUTPUT
+        else:
+            return None
+
+    @staticmethod
+    def parse(parser: BaseParser, short: bool = False) -> "Direction":
+        if (direction := Direction.parse_optional(parser, short)) is None:
+            return parser.raise_error("invalid port direction")
+        return direction
+
+    def print(self, printer: Printer, short: bool = False) -> None:
+        match self:
+            case Direction.INPUT:
+                printer.print("input" if not short else "in")
+            case Direction.OUTPUT:
+                printer.print("output" if not short else "out")
+
+    def is_input_like(self) -> bool:
```

Comment:
Really feels like `is_input` at the moment. Is this meant to change when INOUT is supported? Ive seen a couple of `isInput() &` !isInOut()` in upstream code but not so much `isInput() ||isInOut()` so I’m curious where this is going.

Reply:
Yeah this is the case. I used to go further into the inout implementation and this was useful, notably in the parser/printer to decide on whether it should be an SSA value or not.